pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -268,6 +269,46 @@ protected TopicStatsHelper initialValue() {
     @Getter
     private final ExecutorService orderedExecutor;
 
+    private volatile CloseFutures closeFutures;
+
+    /***
+     * We use 3 futures to prevent a new closing if there is an in-progress deletion or closing.  We make Pulsar return
+     * the in-progress one when it is called the second time.
+     *
+     * The topic closing will be called the below scenarios:
+     * 1. Calling "pulsar-admin topics unload". Relate to {@link CloseFutures#waitDisconnectClients}.
+     * 2. Namespace bundle transfer or unloading.
+     *   a. The unloading topic triggered by unloading namespace bundles will not wait for clients disconnect. Relate
+     *     to {@link CloseFutures#notWaitDisconnectClients}.
+     *   b. The unloading topic triggered by unloading namespace bundles was seperated to two steps when using
+     *     {@link ExtensibleLoadManagerImpl}.
+     *     b-1. step-1: fence the topic on the original Broker, and do not trigger reconnections of clients. Relate
+     *       to {@link CloseFutures#notDisconnectClients}. This step is a half closing.
+     *     b-2. step-2: send the owner broker information to clients and disconnect clients. Relate
+     *       to {@link CloseFutures#notWaitDisconnectClients}.
+     *
+     * The three futures will be setting as the below rule:
+     * Event: Topic close.
+     * - If the first one closing is called by "close and not disconnect clients":
+     *   - {@link CloseFutures#notDisconnectClients} will be initialized as "close and not disconnect clients".
+     *   - {@link CloseFutures#waitDisconnectClients} ang {@link CloseFutures#notWaitDisconnectClients} will be empty,
+     *     the second closing will do a new close after {@link CloseFutures#notDisconnectClients} is completed.
+     * - If the first one closing is called by "close and not wait for clients disconnect":
+     *   - {@link CloseFutures#waitDisconnectClients} will be initialized as "waiting for clients disconnect".
+     *   - {@link CloseFutures#notWaitDisconnectClients} ang {@link CloseFutures#notDisconnectClients} will be
+     *     initialized as "not waiting for clients disconnect" .
+     * - If the first one closing is called by "close and wait for clients disconnect", the three futures will be
+     *   initialized as "waiting for clients disconnect".
+     * Event: Topic delete.
+     *  the three futures will be initialized as "waiting for clients disconnect".
+     */
+    @AllArgsConstructor
+    private class CloseFutures {
+        private final CompletableFuture<Void> waitDisconnectClients;
+        private final CompletableFuture<Void> notWaitDisconnectClients;
```

Comment:
> Why do we need two separate futures here? can't we just combine them, closingOrDeleting?

`notWaitDisconnectClients` is used for bundle unloading, which is faster. `waitDisconnectClients` is used for the Admin API `pulsar-admin topics unload`. 
- `pulsar-admin topics unload` before `bundle unloading`, return `notWaitDisconnectClients`
- `pulsar-admin topics unload` after `bundle unloading`, return `waitDisconnectClients `

Reply:
- unload bundle
  -  the variable `closeFutures.notWaitDisconnectClients` will be initialized as "not wait clients closing"
  -  the variable `closeFutures.waitDisconnectClients` will be initialized as "wait for clients closing"

`admin api unload` will get the `closeFutures.waitDisconnectClients`, and respond successfully after this future is completed. I think this behavior is OK.

Reply:
```java
CompletableFuture<Void> clientCloseFuture = closeWithoutWaitingClientDisconnect
                ? CompletableFuture.completedFuture(null)
                : FutureUtil.waitForAll(futures);
```

Sorry, I am still not convinced. Correct me if I am wrong here. It appears that this logic already makes `clientCloseFuture ` handle both `notWaitDisconnectClients` and `waitDisconnectClients`cases. So, can't we combine these futures?

Reply:
For dedup, I think we can return whatever close future that came first (either notWaitDisconnectClients or waitDisconnectClients)

Reply:
> For dedup, I think we can return whatever the combined close future that came first (either from notWaitDisconnectClients or waitDisconnectClients)

I think the solution implemented in two future instances is better.  Because the bundle unloading process works like this:
1. set the bundle-owned stat to in-active.
2. close topics on the bundle
  2-1. close consumers/producers
3. release the bundle.

The time cost by step `[2,3]` is more, and the time consumers/producers spend on connecting is more. We should better make the process `unloading topic` called by `unload bundle` faster.

Reply:
Case 1: dedup auto unloading during admin unloading

In this case, the current PR returns the faster future(that doesn't wait for producer and consumer close "futures"). I believe the current logic is indeed optimized for this case.


Case 2: dedup admin unloading during auto unloading
However, upon every auto unloading, the current PR also traces the slow future (that does wait for producer and consumer close "futures"), which will increase the overhead of auto-unloading(like I mentioned it in the other comment, about waitForAll(futures)).

I'm fine if we can initialize the slow future on demand, only when required.